pkg/pgclient/config.go

```diff
@@ -135,70 +139,65 @@ func (cfg *Config) GetConnectionStr() string {
 	return cfg.DbUri
 }
 
-func (cfg *Config) GetNumConnections() (min int, max int, numCopiers int, err error) {
-	maxProcs := runtime.GOMAXPROCS(-1)
-	if cfg.WriteConnections < 0 {
-		return 0, 0, 0, fmt.Errorf("invalid number of writer connections %v, cannot be a negative number", cfg.WriteConnections)
+// GetPoolSize returns the max pool size based on the max_connections allowed by database and the defaultFraction.
+// Arg inputPoolSize is the pool size provided in the CLI flag.
+func (cfg *Config) GetPoolSize(poolName string, defaultFraction float64, inputPoolSize int) (int, error) {
+	maxConns, err := cfg.maxConn()
+	if err != nil {
+		return 0, fmt.Errorf("max connections: %w", err)
 	}
 
+	if inputPoolSize == defaultPoolSize {
+		// For the default case, we need to take up defaultFraction of allowed connections.
+		poolSize := float64(maxConns) * defaultFraction
+		if cfg.UsesHA {
+			poolSize /= 2
+		}
+		return int(poolSize), nil
+	}
+
+	// Custom pool size,
+	if inputPoolSize < MinPoolSize {
+		return 0, fmt.Errorf("%s pool size canot be less than %d: received %d", poolName, MinPoolSize, inputPoolSize)
+	} else if inputPoolSize > maxConns {
+		return 0, fmt.Errorf("%s pool size canot be greater than the 'max_connections' allowed by the database", poolName)
+	}
+	if cfg.UsesHA {
+		inputPoolSize /= 2
+	}
+	return inputPoolSize, nil
+}
+
+func (cfg *Config) maxConn() (int, error) {
 	conn, err := pgx.Connect(context.Background(), cfg.GetConnectionStr())
 	if err != nil {
-		return 0, 0, 0, err
+		return 0, err
 	}
 	defer func() { _ = conn.Close(context.Background()) }()
 
-	max = cfg.MaxConnections
-	if max < 1 {
-		var maxStr string
-		row := conn.QueryRow(context.Background(), "SHOW max_connections")
-		err = row.Scan(&maxStr)
-		if err != nil {
-			return 0, 0, 0, err
-		}
-		max, err = strconv.Atoi(maxStr)
-		if err != nil {
-			log.Warn("err", err, "msg", "invalid value from postgres max_connections")
-			max = 100
-		}
-
-		//In HA setups
-		if cfg.UsesHA {
-			max = max / 2
-		}
+	var maxConns int
+	err = conn.QueryRow(context.Background(), "SELECT current_setting('max_connections')::int").Scan(&maxConns)
```

Comment:
Now there is no way to manually configure Promscale to use max connections? It will always use the max from the DB? I'm not sure that's what we want. We can't assume every DB Promscale connects is dedicated to it IMO.

Reply:
Right. Now, we revive the `db.connections-max` flag, which will be used as a check. If total connections (reader max pool size + writer max pool size) > max, then error. So, the user will be forced to either increase the max-conns or decrease the pool sizes.